### PR TITLE
[SPARK-16494] [ML] Upgrade breeze version to 0.12

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

breeze 0.12 has been released for more than half a year, and it brings lots of new features, performance improvement and bug fixes.
One of the biggest features is `LBFGS-B` which is an implementation of `LBFGS` with box constraints and much faster for some special case.
We would like to implement Huber loss function for `LinearRegression` ([SPARK-3181](https://issues.apache.org/jira/browse/SPARK-3181)) and it requires `LBFGS-B` as the optimization solver. So we should bump up the dependent breeze version to 0.12.
For more features, improvements and bug fixes of breeze 0.12, you can refer the following link:
https://groups.google.com/forum/#!topic/scala-breeze/nEeRi_DcY5c
## How was this patch tested?

No new tests, should pass the existing ones.

### CODE REVIEW

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by hvanhovell** [comment]
@yanboliang you need to update the dependency manifests. Run: `./dev/test-dependencies.sh --replace-manifest`

**Comment by hvanhovell** [comment]
NVM...

**Comment by hvanhovell** [comment]
Hmmm it is also pulling in shapeless... did you manually update the manifests?

**Comment by yanboliang** [comment]
@hvanhovell Yes, I updated the dependency manifests manually. I will change to use the scripts you provided. Thanks for your kindly remind!

**Comment by yanboliang** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 1]
```diff
@@ -327,11 +327,6 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
         throw new SparkException(msg)
       }
 
-      if (!state.actuallyConverged) {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Are these changes intentional?

File: dev/deps/spark-deps-hadoop-2.7 [hunk 2]
```diff
@@ -163,6 +163,7 @@ scala-parser-combinators_2.11-1.0.4.jar
 scala-reflect-2.11.8.jar
 scala-xml_2.11-1.0.2.jar
 scalap-2.11.8.jar
+shapeless_2.11-2.0.0.jar
```
Comment (srowen) [inline dev/deps/spark-deps-hadoop-2.7]: We need to check the license of this and possibly update the license files

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Yes, since breeze removed ```state.actuallyConverged``` function and will output this kind of info in version 0.12 automatically.

File: dev/deps/spark-deps-hadoop-2.7 (hunk 2 above)
Comment (yanboliang) [inline dev/deps/spark-deps-hadoop-2.7]: Sure.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala [hunk 3]
```diff
@@ -784,7 +784,13 @@ class DistributedLDAModel private[clustering] (
...
   def topTopicsPerDocument(k: Int): RDD[(Long, Array[Int], Array[Double])] = {
     graph.vertices.filter(LDA.isDocumentVertex).map { case (docID, topicCounts) =>
-      val topIndices = argtopk(topicCounts, k)
+      // TODO: Remove work-around for the breeze bug.
+      // https://github.com/scalanlp/breeze/issues/561
+      val topIndices = if (k == topicCounts.length) {
+        Seq.range(0, k)
+      } else {
+        argtopk(topicCounts, k)
+      }
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: Here we hit the breeze bug (https://github.com/scalanlp/breeze/issues/561), I use work-around here and sent PR to breeze to fix it.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala [hunk 4]
```diff
@@ -508,8 +508,9 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     val weight = rho()
     val N = gammat.rows.toDouble
     val alpha = this.alpha.asBreeze.toDenseVector
-    val logphat: BDM[Double] = sum(LDAUtils.dirichletExpectation(gammat)(::, breeze.linalg.*)) / N
-    val gradf = N * (-LDAUtils.dirichletExpectation(alpha) + logphat.toDenseVector)
+    val logphat: BDV[Double] =
+      sum(LDAUtils.dirichletExpectation(gammat)(::, breeze.linalg.*)).t / N
+    val gradf = N * (-LDAUtils.dirichletExpectation(alpha) + logphat)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: Breeze 0.12 has changed the type of return value of ```DenseMatrix(::, *)``` from ```DenseMatrix``` to ```Transpose[DenseVector]```, so we should make corresponding change.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 5]
```diff
@@ -118,8 +118,8 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
         assert(weights.length == 2)
         val bdvTopicDist = topicDistribution.asBreeze
         val top2Indices = argtopk(bdvTopicDist, 2)
-        assert(top2Indices.toArray === indices)
-        assert(bdvTopicDist(top2Indices).toArray === weights)
+        assert(top2Indices.toSet === indices.toSet)
+        assert(bdvTopicDist(top2Indices).toArray.toSet === weights.toSet)
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: ```argtopk``` does not guarantee the output is sorted. In breeze 0.12, it speed up ```argtopk``` for ```breeze.linalg.DenseVector``` with ```quickSelect``` rather than sorting the vector. So here we should compare the equality of set.

File: mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala [hunk 6]
```diff
@@ -42,7 +43,9 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
     val pca_transform = pca.transform(dataRDD).collect()
     val mat_multiply = mat.multiply(pc).rows.collect()
 
-    assert(pca_transform.toSet === mat_multiply.toSet)
-    assert(pca.explainedVariance === explainedVariance)
+    pca_transform.zip(mat_multiply).foreach { case (calculated: Vector, expected: Vector) =>
+      assert(calculated ~== expected relTol 1e-8)
+    }
+    assert(pca.explainedVariance ~== explainedVariance relTol 1e-8)
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala]: Fix numerical accuracy issue.

File: dev/deps/spark-deps-hadoop-2.7 (hunk 2 above)
Comment (yanboliang) [inline dev/deps/spark-deps-hadoop-2.7]: Checked, Apache license 2.0. So no need to update the license files.

**Comment by yanboliang** [comment]
cc @srowen

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java [hunk 7]
```diff
@@ -107,7 +107,11 @@ public VectorPair call(Tuple2<Vector, Vector> pair) {
       .fit(df);
     List<Row> result = pca.transform(df).select("pca_features", "expected").toJavaRDD().collect();
     for (Row r : result) {
-      Assert.assertEquals(r.get(1), r.get(0));
+      Vector calculatedVector = (Vector)r.get(0);
```
Comment (srowen) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java]: Not sure why these don't fail the style checker, but, space after cats, i++ instead of i ++, and 1.0e-8 for clarity

File: mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala [hunk 8]
```diff
@@ -42,7 +43,9 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
     val pca_transform = pca.transform(dataRDD).collect()
     val mat_multiply = mat.multiply(pc).rows.collect()
 
-    assert(pca_transform.toSet === mat_multiply.toSet)
-    assert(pca.explainedVariance === explainedVariance)
+    pca_transform.zip(mat_multiply).foreach { case (calculated: Vector, expected: Vector) =>
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala]: Not a big deal but do you really need the explicit types?

**Comment by srowen** [comment]
Any other changes that might cause compatibility problems? i doubt it but worth skimming the release notes. This seems OK for 2.1

**Comment by yanboliang** [comment]
@srowen I found no obvious compatibility issues after reading the release notes. If this looks good, please let it get in, since [SPARK-3181](https://issues.apache.org/jira/browse/SPARK-3181) depends upon this.

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java (hunk 7 above)
Comment (srowen) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java]: This needs a rebase and you might run dev/lint-java manually; I think this cast will still fail the style checker?

**Comment by yanboliang** [comment]
Jenkins, test this please.

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java (hunk 7 above)
Comment (yanboliang) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java]: Fix the cast issue, although the style checker does not complain anything. I think we can track the style checker issue in a separate JIRA.

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
HI, this seems to break Scala 2.10 build.

https://amplab.cs.berkeley.edu/jenkins/job/spark-master-compile-maven-scala-2.10/2045/console

**Comment by srowen** [comment]
Rats. I can revert it but weirdly I don't see an actual compile failure there. Just that it failed. I don't think I can retrigger it though. I'll just revert.

**Comment by dongjoon-hyun** [comment]
Oh, sorry. Please stop, @srowen !!

**Comment by dongjoon-hyun** [comment]
It was a flakyness.
The next build succeed!

**Comment by dongjoon-hyun** [comment]
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-master-compile-maven-scala-2.10/2046/

**Comment by srowen** [comment]
OK thanks. Yeah I figured I'd leave it for a minute and I didn't do anything. Not sure what happened there.

**Comment by dongjoon-hyun** [comment]
Great!! Thank you so much!!
Now, it's MY big mistake. Sorry. :)
